blocks/library/categories/index.js

```diff
@@ -0,0 +1,241 @@
+import { getCategories } from './data.js';
+import InspectorControls from '../../inspector-controls';
+import ToggleControl from '../../inspector-controls/toggle-control';
+import BlockDescription from '../../block-description';
+import BlockControls from '../../block-controls';
+import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+
+registerBlockType( 'core/categories', {
+	title: __( 'Categories' ),
+
+	icon: 'list-view',
+
+	category: 'widgets',
+
+	defaultAttributes: {
+		showPostCounts: false,
+		displayAsDropdown: false,
+		showHierarchy: false,
+	},
+
+	getEditWrapperProps( attributes ) {
+		const { align } = attributes;
+		if ( 'left' === align || 'right' === align || 'full' === align ) {
+			return { 'data-align': align };
+		}
+	},
+
+	edit: class extends Component {
+		constructor() {
+			super( ...arguments );
+
+			this.state = {
+				categories: [],
+			};
+
+			this.categoriesRequest = getCategories();
+
+			this.categoriesRequest
+				.then( categories => this.setState( { categories } ) );
+
+			this.toggleDisplayAsDropdown = this.toggleDisplayAsDropdown.bind( this );
+			this.toggleShowPostCounts = this.toggleShowPostCounts.bind( this );
+			this.toggleShowHierarchy = this.toggleShowHierarchy.bind( this );
+		}
+
+		toggleDisplayAsDropdown() {
+			const { displayAsDropdown } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { displayAsDropdown: ! displayAsDropdown } );
+		}
+
+		toggleShowPostCounts() {
+			const { showPostCounts } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { showPostCounts: ! showPostCounts } );
+		}
+
+		toggleShowHierarchy() {
+			const { showHierarchy } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { showHierarchy: ! showHierarchy } );
+		}
+
+		getCategories( parentId = null ) {
+			const { categories } = this.state;
+			if ( ! categories.length ) {
+				return [];
+			}
+
+			if ( parentId === null ) {
+				return categories;
+			}
+
+			return categories.filter( category => category.parent === parentId );
+		}
+
+		getCategoryListClassName( level ) {
+			const { className } = this.props;
+			return `${ className }__list ${ className }__list-level-${ level }`;
+		}
+
+		renderCategoryList() {
+			const { showHierarchy } = this.props.attributes;
+			const parentId = showHierarchy ? 0 : null;
+			const categories = this.getCategories( parentId );
+
+			return (
+				<ul className={ this.getCategoryListClassName( 0 ) }>
+					{ categories.map( category => this.renderCategoryListItem( category, 0 ) ) }
+				</ul>
+			);
+		}
+
+		renderCategoryListItem( category, level ) {
+			const { showHierarchy, showPostCounts } = this.props.attributes;
+			const childCategories = this.getCategories( category.id );
+
+			return (
+				<li key={ category.id }>
+					<a href={ category.link } target="_blank">{ category.name.trim() || __( '(Untitled)' ) }</a>
```

Comment:
Curious since I've dealt with issues before: How well does this handle entities, like an ampersand `&` ? Might need to do some unescaping.